rust/kernel/net/rtnl.rs

```diff
@@ -0,0 +1,148 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Rtnl Link Operations.
+//!
+//! C header: [`include/linux/rtnetlink.h`](../../../../include/linux/rtnetlink.h)
+
+use core::ptr;
+
+use crate::bindings;
+use crate::error::{Error, Result};
+
+use super::device::{NetDevice, NetDeviceAdapter};
+
+use crate::types::{SavedAsPointer, SavedAsPointerMut};
+
+// TODO: inner bool, to allow other unlock mechanism?
+/// Lock, acquired via [`bindings::rtnl_lock`].
+#[must_use = "the rtnl unlocks immediately when the guard is unused"]
+pub struct RtnlLock {
+    _private: (),
+}
+
+impl RtnlLock {
+    /// Creat a new lock via [`bindings::rtnl_lock`].
+    pub fn lock() -> Self {
+        // SAFETY: C function without parameters
+        unsafe { bindings::rtnl_lock() };
+
+        Self { _private: () }
+    }
+}
+
+impl Drop for RtnlLock {
+    fn drop(&mut self) {
+        // SAFETY: C function without parameters
+        unsafe { bindings::rtnl_unlock() };
+    }
+}
+
+#[doc(hidden)]
+/// Empty [`bindings::rtnl_link_ops`] table.
+pub const RTNL_LINK_OPS_EMPTY: bindings::rtnl_link_ops = bindings::rtnl_link_ops {
+    list: bindings::list_head {
+        next: ptr::null::<bindings::list_head>() as *mut bindings::list_head,
+        prev: ptr::null::<bindings::list_head>() as *mut bindings::list_head,
+    },
+    kind: ptr::null::<i8>(),
+    priv_size: 0,
+    setup: None,
+    maxtype: 0,
+    policy: ptr::null::<bindings::nla_policy>(),
+    validate: None,
+    newlink: None,
+    changelink: None,
+    dellink: None,
+    get_size: None,
+    fill_info: None,
+    get_xstats_size: None,
+    fill_xstats: None,
+    get_num_tx_queues: None,
+    get_num_rx_queues: None,
+    slave_maxtype: 0,
+    slave_policy: ptr::null::<bindings::nla_policy>(),
+    slave_changelink: None,
+    get_slave_size: None,
+    fill_slave_info: None,
+    get_link_net: None,
+    get_linkxstats_size: None,
+    fill_linkxstats: None,
+};
+
+/// Transparent wrapper for [`bindings::rtnl_link_ops`] for the macro [`kernel::rtnl_link_ops`].
+#[repr(transparent)]
+pub struct RtnlLinkOps(pub bindings::rtnl_link_ops);
+
+unsafe impl Sync for RtnlLinkOps {}
+
+impl RtnlLinkOps {
+    /// Register this op table with kernel.
+    pub fn register(&self) -> Result {
+        // SAFETY: ptr of self is valid if self is valid
+        let ret = unsafe {
+            let ptr = self.get_pointer();
+
+            bindings::rtnl_link_register(ptr as *mut bindings::rtnl_link_ops)
+        };
+
+        if ret != 0 {
+            Err(Error::from_kernel_errno(ret))
+        } else {
+            Ok(())
+        }
+    }
+
+    /// Get a pointer to this struct.
+    pub unsafe fn get_pointer(&self) -> *const bindings::rtnl_link_ops {
+        &self.0 as *const _ as *const bindings::rtnl_link_ops
```

Comment:
the `as *const _` seems redundant

Reply:
Yes, but I think rust wants them, as this is not directly the right type for the view of rust

Reply:
I think in this case you don't need it is since you're turning a `&rtnl_link_ops` into a `*const rtnl_link_ops`. I agree that the other ones are necessary.

(I could of course be wrong about that and it's ultimately not that important)

Reply:
Could be that I missed .0 befor. And then you need this. Is it is valid, but rust doesn't know that.

Reply:
That was probably the case, you can only do `&T as *const T` but not `&U as *const T`, even if `U` only contains a `T`.

Reply:
Yes, as U has repr Transparent. It would also be valid, but rust isn't able to check that. That's the reason to go over *const _